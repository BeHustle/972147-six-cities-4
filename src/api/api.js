import axios from 'axios';

const URL = `https://4.react.pages.academy/six-cities`;
const TIMEOUT = 5000;

export const createAPI = (dispatch = null) => {
  const api = axios.create({
    baseURL: URL,
    timeout: TIMEOUT,
    withCredentials: true
  });

  const onSuccess = (response) => response;
  const onFail = ({response}) => {
    dispatch();
    throw response;
  };

  api.interceptors.response.use(onSuccess, onFail);
  return api;
};
