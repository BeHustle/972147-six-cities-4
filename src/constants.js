const CARD_TYPES = [
  `Apartment`,
  `Private room`
];

const Cities = {
  AMSTERDAM: [52.38333, 4.9]
};

const MAP_ZOOM = 12;

const ICON_SIZE = [30, 30];

const ICON_PATH = `img/pin.svg`;

export {CARD_TYPES, Cities, MAP_ZOOM, ICON_SIZE, ICON_PATH};
