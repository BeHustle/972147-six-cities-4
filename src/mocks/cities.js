export const cities = [
  {
    id: 1,
    name: `Amsterdam`,
    coordinates: [52.38333, 4.9]
  },
  {
    id: 2,
    name: `Paris`,
    coordinates: [52.38333, 4.9]
  },
  {
    id: 3,
    name: `Cologne`,
    coordinates: [52.38333, 4.9]
  },
  {
    id: 4,
    name: `Brussels`,
    coordinates: [52.38333, 4.9]
  },
  {
    id: 5,
    name: `Hamburg`,
    coordinates: [52.38333, 4.9]
  },
  {
    id: 6,
    name: `Dusseldorf`,
    coordinates: [52.38333, 4.9]
  },
];
