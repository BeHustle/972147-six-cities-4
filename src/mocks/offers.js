export const offers = [
  {
    id: 1,
    name: `Beautiful & luxurious apartment at great location`,
    price: 132,
    image: `img/apartment-01.jpg`,
    type: `Private room`,
    isPremium: false,
    inBookmarks: true,
    rating: 4
  },
  {
    id: 2,
    name: `Wood and stone place`,
    price: 254,
    image: `img/apartment-02.jpg`,
    type: `Apartment`,
    isPremium: false,
    inBookmarks: false,
    rating: 5
  },
  {
    id: 3,
    name: `Canal View Prinsengracht`,
    price: 178,
    image: `img/apartment-03.jpg`,
    type: `Private room`,
    isPremium: true,
    inBookmarks: true,
    rating: 3
  },
  {
    id: 4,
    name: `Nice, cozy, warm big bed apartment`,
    price: 995,
    image: `img/room.jpg`,
    type: `Apartment`,
    isPremium: true,
    inBookmarks: false,
    rating: 1
  },
  {
    id: 5,
    name: `Beautiful & luxurious apartment at great location`,
    price: 80,
    image: `img/apartment-01.jpg`,
    type: `Private room`,
    isPremium: false,
    inBookmarks: true,
    rating: 2
  },
];
