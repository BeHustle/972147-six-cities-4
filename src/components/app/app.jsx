import React from 'react';
import {BrowserRouter, Switch, Route} from 'react-router-dom';
import {getAppStatus} from '../../reducer/app/app.selectors.js';
import CardDetail from '../card-detail/card-detail.jsx';
import Main from '../main/main.jsx';
import {Screen, AppStatus} from '../../constants.js';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import SingIn from '../sign-in/sign-in.jsx';

class App extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      screen: Screen.MAIN,
      offerId: null,
    };
    this._handleCardClick = this._handleCardClick.bind(this);
    this._handleLoginClick = this._handleLoginClick.bind(this);
  }

  _handleCardClick(id) {
    this.setState({
      screen: Screen.OFFER,
      offerId: id
    });
  }

  _handleLoginClick() {
    this.setState({
      screen: Screen.LOGIN,
    });
  }

  _renderApp() {
    switch (this.props.status) {
      case AppStatus.LOADING:
        return <h1>Загрузка...</h1>;
      case AppStatus.FAIL_LOAD:
        return <h1>Произошла ошибка при загрузке приложения. Попробуйте <a href="/">обновить страницу</a></h1>;
      case AppStatus.SUCCESS_LOAD:
        return this._renderScreen();
      default:
        return ``;
    }
  }

  _renderScreen() {
    switch (this.state.screen) {
      case Screen.MAIN:
        return <Main onCardTitleClick={this._handleCardClick} onSignInClick={this._handleLoginClick} />;
      case Screen.OFFER:
        return <CardDetail onCardTitleClick={this._handleCardClick} offerId={this.state.offerId} onSignInClick={this._handleLoginClick}/>;
      case Screen.LOGIN:
        return <SingIn />;
      default:
        return null;
    }
  }

  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/offer">
            <CardDetail onCardTitleClick={this._handleCardClick} offerId={1} />
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

App.propTypes = {
  status: PropTypes.oneOf(Object.values(AppStatus))
};

const mapStateToProps = (state) => ({
  status: getAppStatus(state)
});

export default connect(mapStateToProps)(App);
