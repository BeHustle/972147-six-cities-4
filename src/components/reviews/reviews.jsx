import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {AuthStatus} from '../../constants.js';
import {Operation as DataOperation} from '../../reducer/data/data.reducer.js';
import {getReviews} from '../../reducer/data/data.selectors.js';
import {getAuthStatus} from '../../reducer/user/user.selectors.js';
import AddReview from '../add-review/add-review.jsx';
import Review from '../review/review.jsx';

class Reviews extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onReviewsMount, offerId} = this.props;
    onReviewsMount(offerId);
  }

  render() {
    const {reviews, authStatus, offerId} = this.props;
    return <section className="property__reviews reviews">
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
      <ul className="reviews__list">
        {reviews.map((review) =>
          <Review key={review.id} review={review}/>,
        )}
      </ul>
      {authStatus === AuthStatus.AUTH && <AddReview offerId={offerId} />}
    </section>;
  }
}

Reviews.propTypes = {
  offerId: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.exact({
    id: PropTypes.number.isRequired,
    author: PropTypes.exact({
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string,
    }),
    text: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    date: PropTypes.instanceOf(Date),
  })),
  onReviewsMount: PropTypes.func.isRequired,
  authStatus: PropTypes.oneOf(Object.values(AuthStatus)),
};

const mapDispatchToProps = (dispatch) => ({
  onReviewsMount(offerId) {
    dispatch(DataOperation.loadReviews(offerId));
  },
});

const mapStateToProps = (state) => ({
  reviews: getReviews(state),
  authStatus: getAuthStatus(state)
});

export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
