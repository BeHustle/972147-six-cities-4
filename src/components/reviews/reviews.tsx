import * as React from 'react';
import {connect} from 'react-redux';
import {AuthStatus} from '../../constants';
import {Operation as DataOperation} from '../../reducer/data/data.reducer';
import {getReviews} from '../../reducer/data/data.selectors';
import {getAuthStatus} from '../../reducer/user/user.selectors';
import AddReview from '../add-review/add-review';
import Review from '../review/review';
import {ReviewInterface} from "../../types";

interface Props {
  offerId: number;
  onReviewsMount: (offerId: number) => void;
  reviews: Array<ReviewInterface>;
  authStatus: string;
}

class Reviews extends React.PureComponent<Props, {}> {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {onReviewsMount, offerId} = this.props;
    onReviewsMount(offerId);
  }

  render() {
    const {reviews, authStatus, offerId} = this.props;
    return <section className="property__reviews reviews">
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
      <ul className="reviews__list">
        {reviews.map((review) =>
          <Review key={review.id} review={review}/>,
        )}
      </ul>
      {authStatus === AuthStatus.AUTH && <AddReview offerId={offerId} />}
    </section>;
  }
}

const mapDispatchToProps = (dispatch) => ({
  onReviewsMount(offerId) {
    dispatch(DataOperation.loadReviews(offerId));
  },
});

const mapStateToProps = (state) => ({
  reviews: getReviews(state),
  authStatus: getAuthStatus(state)
});

export {Reviews};

export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
