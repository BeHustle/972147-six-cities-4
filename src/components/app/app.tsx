import * as React from 'react';
import {Router, Switch, Route, Link} from 'react-router-dom';
import {getAppStatus} from '../../reducer/app/app.selectors';
import {getAuthStatus} from '../../reducer/user/user.selectors';
import AppRoute from '../../routes';
import CardDetail from '../card-detail/card-detail';
import Favorites from '../favorites/favorites';
import Main from '../main/main';
import {AppStatus, AuthStatus} from '../../constants';
import {connect} from 'react-redux';
import NotFound from '../not-found/not-found';
import PublicRoute from '../public-route/public-route';
import SingIn from '../sign-in/sign-in';
import {history} from '../../history';
import PrivateRoute from '../private-route/private-route';
import {Operation as DataOperation} from '../../reducer/data/data.reducer';

interface Props {
  appStatus: string;
  authStatus: string;
  onFavoriteClick: (offerId: number, status: string) => void;
}

class App extends React.PureComponent<Props, {}> {
  constructor(props) {
    super(props);
    this._handleCardClick = this._handleCardClick.bind(this);
    this._handleFavoriteClick = this._handleFavoriteClick.bind(this);
  }

  _handleCardClick(id) {
    history.push(`${AppRoute.OFFER}/${id}`);
  }

  _handleFavoriteClick(offerId, status) {
    if (this.props.authStatus === AuthStatus.AUTH) {
      this.props.onFavoriteClick(offerId, status);
    } else {
      history.push(AppRoute.LOGIN);
    }
  }

  render() {
    switch (this.props.appStatus) {
      case AppStatus.LOADING:
        return <h1>Loading...</h1>;
      case AppStatus.FAIL_LOAD:
        return <h1>Error while loading. Try <Link to={AppRoute.MAIN}>refresh page</Link></h1>;
      case AppStatus.SUCCESS_LOAD:
        return (
          <Router history={history}>
            <Switch>
              <Route exact path={AppRoute.MAIN}>
                <Main onCardTitleClick={this._handleCardClick} onFavoriteClick={this._handleFavoriteClick} />
              </Route>
              <Route
                exact
                path={`${AppRoute.OFFER}/:offerId`}
                render={(props) =>
                  <CardDetail
                    offerId={parseInt(props.match.params.offerId, 10)}
                    onCardTitleClick={this._handleCardClick}
                    onFavoriteClick={this._handleFavoriteClick}
                  />
                }
              />
              <PublicRoute exact path={AppRoute.LOGIN}>
                <SingIn />
              </PublicRoute>
              <PrivateRoute exact path={AppRoute.FAVORITES}>
                <Favorites
                  onFavoriteClick={this._handleFavoriteClick}
                  onCardTitleClick={this._handleCardClick}
                />
              </PrivateRoute>
              <Route>
                <NotFound />
              </Route>
            </Switch>
          </Router>
        );
      default:
        return ``;
    }
  }
}

const mapStateToProps = (state) => ({
  appStatus: getAppStatus(state),
  authStatus: getAuthStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  onFavoriteClick(offerId, status) {
    dispatch(DataOperation.addFavoriteOffer({
      hotelId: offerId,
      status
    }));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
