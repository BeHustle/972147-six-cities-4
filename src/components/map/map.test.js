import MockAdapter from 'axios-mock-adapter';
import * as React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import {applyMiddleware, createStore} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import {createAPI} from '../../api/api';
import {AppStatus, AuthStatus, CardType} from '../../constants';
import {setActiveCity, setAppStatus} from '../../reducer/app/app.reducer';
import {setCities, setNearbyOffers, setOffers, setReviews} from '../../reducer/data/data.reducer';
import reducer from '../../reducer/reducer';
import {setAuthStatus, setUserInfo} from '../../reducer/user/user.reducer';
import {reviews, serverReviews} from '../../test-mocks/reviews';
import Map from './map.tsx';
import {cities} from '../../test-mocks/cities';
import {offers, serverOffers} from '../../test-mocks/offers';
import {serverUserInfo, userInfo} from '../../test-mocks/user';

jest.mock(`../map/mapx`, () => `map`);

const api = createAPI();
const apiMock = new MockAdapter(api);

apiMock
  .onGet(`/hotels`)
  .reply(200, serverOffers);

apiMock
  .onGet(`/comments/1`)
  .reply(200, serverReviews);

apiMock
  .onGet(`/hotels/1/nearby`)
  .reply(200, serverOffers);

apiMock
  .onGet(`/login`)
  .reply(200, serverUserInfo);

const store = createStore(
    reducer,
    composeWithDevTools(
        applyMiddleware(thunk.withExtraArgument(api)),
    ),
);

store.dispatch(setOffers(offers));
store.dispatch(setReviews(reviews));
store.dispatch(setNearbyOffers(offers));
store.dispatch(setCities(cities));
store.dispatch(setActiveCity(cities[0]));
store.dispatch(setAppStatus(AppStatus.SUCCESS_LOAD));
store.dispatch(setUserInfo(userInfo));
store.dispatch(setAuthStatus(AuthStatus.AUTH));

it(`Render Map`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <Map
            type={CardType.MAIN}
            offers={offers}
            coordinates={cities[0].coordinates}
          />
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
