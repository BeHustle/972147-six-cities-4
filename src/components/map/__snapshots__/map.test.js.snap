// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Map 1`] = `
<map
  coordinates={
    Array [
      52.38333,
      4.9,
    ]
  }
  offers={
    Array [
      Object {
        "author": Object {
          "avatar": "img/avatar-angelina.jpg",
          "isSuper": true,
          "name": "Angelina",
        },
        "cityId": 1,
        "coordinates": Array [
          52.3909553943508,
          4.85309666406198,
        ],
        "facilities": Array [
          "Wi-Fi",
          "Washing machine",
          "Towels",
          "Heating",
          "Coffee machine",
          "Baby seat",
          "Kitchen",
          "Dishwasher",
          "Cabel TV",
          "Fridge",
        ],
        "guests": "Max 4 adults",
        "id": 1,
        "image": "img/apartment-01.jpg",
        "images": Array [
          "img/room.jpg",
          "img/apartment-01.jpg",
          "img/apartment-02.jpg",
          "img/apartment-03.jpg",
          "img/studio-01.jpg",
          "img/apartment-01.jpg",
        ],
        "inBookmarks": true,
        "isPremium": true,
        "name": "Beautiful & luxurious apartment at great locatccion",
        "price": 132,
        "rating": 4.1,
        "rooms": "3 Bedrooms",
        "text": Array [
          "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.",
          "An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.",
        ],
        "type": "hotel",
        "zoom": 10,
      },
      Object {
        "author": Object {
          "avatar": "img/avatar-max.jpg",
          "isSuper": true,
          "name": "Max",
        },
        "cityId": 1,
        "coordinates": Array [
          52.369553943508,
          4.85309666406198,
        ],
        "facilities": Array [
          "Wi-Fi",
          "Washing machine",
          "Towels",
        ],
        "guests": "Max 2 adults",
        "id": 2,
        "image": "img/apartment-02.jpg",
        "images": Array [
          "img/room.jpg",
          "img/apartment-01.jpg",
          "img/apartment-02.jpg",
          "img/apartment-03.jpg",
          "img/studio-01.jpg",
          "img/apartment-01.jpg",
        ],
        "inBookmarks": false,
        "isPremium": false,
        "name": "Wood and stone place",
        "price": 254,
        "rating": 5,
        "rooms": "1 Bedrooms",
        "text": Array [
          "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.",
          "An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.",
        ],
        "type": "apartment",
        "zoom": 10,
      },
      Object {
        "author": Object {
          "avatar": "",
          "isSuper": true,
          "name": "Alex",
        },
        "cityId": 2,
        "coordinates": Array [
          48.85875194,
          2.33926391,
        ],
        "facilities": Array [
          "Wi-Fi",
          "Heating",
          "Coffee machine",
          "Baby seat",
          "Kitchen",
        ],
        "guests": "Max 2 adults and 1 children",
        "id": 3,
        "image": "img/apartment-03.jpg",
        "images": Array [
          "img/room.jpg",
          "img/apartment-01.jpg",
          "img/apartment-02.jpg",
          "img/apartment-03.jpg",
          "img/studio-01.jpg",
          "img/apartment-01.jpg",
        ],
        "inBookmarks": true,
        "isPremium": true,
        "name": "Canal View Prinsengracht",
        "price": 178,
        "rating": 3,
        "rooms": "2 Bedrooms",
        "text": Array [
          "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.",
          "An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.",
        ],
        "type": "room",
        "zoom": 10,
      },
      Object {
        "author": Object {
          "avatar": "",
          "isSuper": true,
          "name": "Steven",
        },
        "cityId": 3,
        "coordinates": Array [
          50.938361,
          6.959974,
        ],
        "facilities": Array [
          "Wi-Fi",
          "Kitchen",
          "Dishwasher",
          "Cabel TV",
          "Fridge",
        ],
        "guests": "Max 3 adults and 2 children",
        "id": 4,
        "image": "img/room.jpg",
        "images": Array [
          "img/room.jpg",
          "img/apartment-01.jpg",
          "img/apartment-02.jpg",
          "img/apartment-03.jpg",
          "img/studio-01.jpg",
          "img/apartment-01.jpg",
        ],
        "inBookmarks": false,
        "isPremium": true,
        "name": "Nice, cozy, warm big bed apartment",
        "price": 995,
        "rating": 1,
        "rooms": "5 Bedrooms",
        "text": Array [
          "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
      The building is green and from 18th century.",
          "An independent House, strategically located between Rembrand Square and National Opera, but where the
      bustle of the city comes to rest in this alley flowery and colorful.",
        ],
        "type": "house",
        "zoom": 10,
      },
    ]
  }
  type="main"
/>
`;
